File: js-jasmine/spec/gilded_rose_spec.js
let { Shop, Item } = require("../src/gilded_rose.js");
describe("Gilded Rose", function () {

    /* 0
February's fool
  */

  it("should foo", function() {
    const gildedRose = new Shop([ new Item("foo", 0, 0) ]);
    const items = gildedRose.updateQuality();
    expect(items[0].name).toEqual("fixme");
  });
  
  /* 1
La qualité baisse de 1 pour les articles standards.
*/

  it("standard items decrease in quality by 1", function () {
    let gildedRose = new Shop([new Item("Standard Item", 10, 20)]);
    let items = gildedRose.updateQuality();
    expect(items[0].quality).toEqual(19);
  });
});

/* 2
Une fois la date de péremption passée, la qualité baisse de 2 pour les articles standards.
*/

it("Standard items decrease in quality by 2 after the sell date has passed", function() {
  const gildedRose = new Shop([new Item("Standard Item", 0, 10)]);
  const items = gildedRose.updateQuality();
  expect(items[0].quality).toEqual(8);
});


/* 3
La qualité d'un article n'est jamais négative.
*/

it("An item's quality is never negative", function() {
  const gildedRose = new Shop([new Item("Standard Item", 5, 0)]);
  const items = gildedRose.updateQuality();
  expect(items[0].quality).toEqual(0);
});

/* 4
"Aged Brie" augmente en qualité au fil du temps.
*/

it("Aged Brie increases in quality the older it gets", function() {
  const gildedRose = new Shop([new Item("Aged Brie", 2, 0)]);
  const items = gildedRose.updateQuality();
  expect(items[0].quality).toEqual(1);
});


/* 5
La qualité d'un article n'est jamais supérieure à 50, sauf "Sulfuras".
*/

it("An item's quality is never more than 50, except Sulfuras", function() {
  const gildedRose = new Shop([new Item("Aged Brie", 2, 50)]);
  const items = gildedRose.updateQuality();
  expect(items[0].quality).toEqual(50);
});


/* 6
"Sulfuras" ne perd jamais en qualité et n'a pas de date de péremption.
*/

it("Sulfuras never decreases in quality and does not have a sell-by date", function() {
  const gildedRose = new Shop([new Item("Sulfuras, Hand of Ragnaros", null, 80)]);
  const items = gildedRose.updateQuality();
  expect(items[0].quality).toEqual(80);
});


/* 7
"Backstage passes" augmente en qualité comme "Aged Brie", mais la qualité tombe à 0 après le concert.
*/

it("Backstage passes increase in quality like Aged Brie but drop to 0 after the concert", function() {
  const gildedRose = new Shop([new Item("Backstage passes to a TAFKAL80ETC concert", 0, 20)]);
  const items = gildedRose.updateQuality();
  expect(items[0].quality).toEqual(0);
});


/* 8
"Conjured" items se dégradent en qualité deux fois plus vite que les articles normaux.
*/

it("Conjured items degrade in quality twice as fast as normal items", function() {
  const gildedRose = new Shop([new Item("Conjured Mana Cake", 3, 6)]);
  const items = gildedRose.updateQuality();
  expect(items[0].quality).toEqual(4); 
});

File: ruby/gilded_rose_spec.rb
require_relative './gilded_rose.rb'

describe GildedRose do
  describe "#update_quality" do
    it "does not change the name" do
      items = [Item.new("foo", 10, 20)]
      GildedRose.new(items).update_quality()
      expect(items[0].name).to eq "foo"
    end

    it "decreases quality by 1 for standard items" do
      items = [Item.new("Standard Item", 10, 20)]
      GildedRose.new(items).update_quality()
      expect(items[0].quality).to eq 19
    end

    it "decreases quality twice as fast after the sell date has passed" do
      items = [Item.new("Standard Item", 0, 20)]
      GildedRose.new(items).update_quality()
      expect(items[0].quality).to eq 18
    end

    it "never allows negative quality" do
      items = [Item.new("Standard Item", 10, 0)]
      GildedRose.new(items).update_quality()
      expect(items[0].quality).to be >= 0
    end

    it "increases quality of 'Aged Brie' over time" do
      items = [Item.new("Aged Brie", 2, 0)]
      GildedRose.new(items).update_quality()
      expect(items[0].quality).to eq 1
    end

    it "'Sulfuras' never decreases in quality and its sell_in date does not change" do
      items = [Item.new("Sulfuras, Hand of Ragnaros", 10, 80)]
      GildedRose.new(items).update_quality()
      expect(items[0].quality).to eq 80
      expect(items[0].sell_in).to eq 10
    end

    context "'Backstage passes' quality" do
      it "increases by 2 when there are 10 days or less" do
        items = [Item.new("Backstage passes to a TAFKAL80ETC concert", 10, 20)]
        GildedRose.new(items).update_quality()
        expect(items[0].quality).to eq 22
      end

      it "increases by 3 when there are 5 days or less" do
        items = [Item.new("Backstage passes to a TAFKAL80ETC concert", 5, 20)]
        GildedRose.new(items).update_quality()
        expect(items[0].quality).to eq 23
      end

      it "drops to 0 after the concert" do
        items = [Item.new("Backstage passes to a TAFKAL80ETC concert", 0, 20)]
        GildedRose.new(items).update_quality()
        expect(items[0].quality).to eq 0
      end
    end

    it "'Conjured' items decrease in quality twice as fast as normal items" do
      items = [Item.new("Conjured Mana Cake", 10, 20)]
      GildedRose.new(items).update_quality()
      expect(items[0].quality).to eq 18
    end
  end
end
